src/java/org/apache/cassandra/transport/messages/ExecuteMessage.java

```diff
@@ -135,20 +134,16 @@ protected Message.Response execute(QueryState state, long queryStartNanoTime, bo
             if (prepared == null)
                 throw new PreparedQueryNotFoundException(statementId);
 
-            if (!prepared.fullyQualified
-                && !Objects.equals(state.getClientState().getRawKeyspace(), prepared.keyspace)
-                // We can not reliably detect inconsistencies for batches yet
-                && !(prepared.statement instanceof BatchStatement)
-            )
+            if (!prepared.fullyQualified && prepared.statement.eligibleAsPreparedStatement() && !Objects.equals(state.getClientState().getRawKeyspace(), prepared.keyspace))
```

Comment:
Nit: I would add a comment to explain the logic as the last part is not obvious.